modules/cart/commerce_cart.module

```diff
@@ -103,6 +104,20 @@ function commerce_cart_commerce_order_delete(OrderInterface $order) {
 }
 
 /**
+ * Implements hook_ENTITY_TYPE_update().
+ */
+function commerce_cart_commerce_order_presave(OrderInterface $order) {
```

Comment:
It makes sense for this to be done for both carts and admin orders, but didn't you catch this bug by testing checkout? There's this in CheckoutFlowBase which should have handled it:
      if ($next_step_id == 'complete') {
        // @todo Do this earlier in the process.
        $this->order->cart = FALSE;
        // Place the order.
        $transition = $this->order->getState()->getWorkflow()->getTransition('place');
        $this->order->getState()->applyTransition($transition);
        // Notify other modules.
        $dispatch_checkout_complete = TRUE;
      }